exporter/googlecloudpubsubexporter/exporter.go

```diff
@@ -16,54 +16,134 @@ package googlecloudpubsubexporter
 
 import (
 	"context"
+	"fmt"
 
+	pubsub "cloud.google.com/go/pubsub/apiv1"
+	"github.com/google/uuid"
 	"go.opentelemetry.io/collector/component"
-	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/otlp"
 	"go.opentelemetry.io/collector/model/pdata"
 	"go.uber.org/zap"
+	"google.golang.org/api/option"
+	pubsubpb "google.golang.org/genproto/googleapis/pubsub/v1"
+	"google.golang.org/grpc"
 )
 
 const name = "googlecloudpubsub"
 
 type pubsubExporter struct {
-	instanceName string
-	logger       *zap.Logger
-
-	topicName string
-
-	//
-	userAgent string
-	ceSource  string
-	config    *Config
-	//
+	instanceName     string
+	logger           *zap.Logger
+	topicName        string
+	client           *pubsub.PublisherClient
+	cancel           context.CancelFunc
+	userAgent        string
+	ceSource         string
+	config           *Config
+	tracesMarshaler  pdata.TracesMarshaler
+	metricsMarshaler pdata.MetricsMarshaler
+	logsMarshaler    pdata.LogsMarshaler
 }
 
 func (*pubsubExporter) Name() string {
 	return name
 }
 
+type Encoding int
+
+const (
+	OtlpProtoTrace  = iota
+	OtlpProtoMetric = iota
+	OtlpProtoLog    = iota
+)
+
 func (ex *pubsubExporter) start(ctx context.Context, _ component.Host) error {
+	ctx, ex.cancel = context.WithCancel(ctx)
+
+	if ex.client == nil {
+		copts := ex.generateClientOptions()
+		client, err := pubsub.NewPublisherClient(ctx, copts...)
+		if err != nil {
+			return fmt.Errorf("failed creating the gRPC client to Pubsub: %w", err)
+		}
+
+		ex.client = client
+	}
+	ex.tracesMarshaler = otlp.NewProtobufTracesMarshaler()
+	ex.metricsMarshaler = otlp.NewProtobufMetricsMarshaler()
+	ex.logsMarshaler = otlp.NewProtobufLogsMarshaler()
 	return nil
 }
 
 func (ex *pubsubExporter) shutdown(context.Context) error {
+	if ex.client != nil {
+		ex.client.Close()
+		ex.client = nil
+	}
 	return nil
 }
 
-func (ex *pubsubExporter) Capabilities() consumer.Capabilities {
-	return consumer.Capabilities{
-		MutatesData: false,
+func (ex *pubsubExporter) generateClientOptions() []option.ClientOption {
+	var copts []option.ClientOption
```

Comment:
```suggestion
func (ex *pubsubExporter) generateClientOptions() (copts []option.ClientOption)
```

Saves you needed to declare as a var.